Provide an error message in slack output when NagBot can't stop / terminate resources

alizbay94 opened the pull request:
With the changes in this PR, an error message will be included in the message sent to slack when nagbot runs into trouble stopping / terminating resources. 

In each of the resource's terminate and stop methods, `True` is returned if the resource is terminated, and the error message is returned if an error occurs during termination / stopping. That error message, and the resource it pertains to, is included in the final slack message.

The pytests have been updated to reflect the new changes.

alizbay94 committed:
Include an error message in the slack output when nagbot is unable to terminate or stop instances

alizbay94 committed:
Remove forgotten comment

sejal-raval commented on app/ami.py:
```diff
@@ -134,8 +133,8 @@ def terminate_resource(self, dryrun: bool) -> bool:
                     print(f'Failure when calling snapshot.delete(): {str(e)}')
-                    snapshots_deleted = False  # set to False and continue attempting to delete remaining Snapshots
-        return snapshots_deleted
+                    return e
```
The previous logic was attempting to delete the remaining Snapshots after the exception. With the new logic, if any error occurs, the process will not continue attempting.

It's a design decision for this CRAB - whether we want to stop the process when an exception occurs or to continue attempting to delete the remaining resources. I think the process should be uniform for all resources i.e. instances, ami, snapshot, and volume.

alizbay94 replied:
I changed the logic since I assume if there is an issue deleting one of the snapshots composing the AMI, the other snapshots would have a similar issue as well, but to be sure, I can tweak the logic to be similar to the previous logic so the process will continue iterating through the remaining snapshots if an error is encountered.

alizbay94 replied:
Just pushed changes that allow the process to continue even if an error is raised, if there are any errors they will all be included in the final slack message. Verified the pytests pass with those changes 👍🏼 .

sejal-raval commented on app/instance.py:
```diff
@@ -133,7 +133,7 @@ def terminate_resource(self, dryrun: bool) -> bool:
         except Exception as e:
             print(f'Failure when calling terminate_instances: {str(e)}')
-            return False
+            return e
```
method return indicates `bool` whereas `e` is `Exception`

alizbay94 replied:
Made a change to remove the return type from the method, and returned the exception as a `string` rather than an `Exception'.

sejal-raval commented on app/nagbot.py:
```diff
@@ -116,11 +116,15 @@ def execute_internal(channel, dryrun):
+                    response = r.terminate_resource(dryrun=dryrun)
+                    if response is not True:
+                        message = message + f"Error when attempting to terminate {r.make_resource_summary()}:" \
+                                            f" {response}\n"
```
`f" {str(response)}\n"` can be more descriptive.

alizbay94 replied:
`f" {response}\n"` is prefixed with `Error when attempting to terminate {r.make_resource_summary()}:`, I think that should be enough information to diagnose the error (unless you're referencing the fact that response would be of type `Exception` so it needs to be typecasted to string; if that's the case, I changed the return type of `terminate_resource()` from `Exception` to `String` when a termination error occurs, so we won't need to include `str()` around `response` any more as in `str(response)` since response is now a string in that case.

alizbay94 committed:
Continue attempting to delete Snapshots belonging to AMI even if one of the snapshot delete operations raise an error

sejal-raval commented on app/nagbot.py:
```diff
@@ -116,11 +116,15 @@ def execute_internal(channel, dryrun):
-                        f'{util.money_to_string(r.monthly_price)}, Contact={contact}\n'
-                    r.terminate_resource(dryrun=dryrun)
+                    response = r.terminate_resource(dryrun=dryrun)
+                    if response is not True:
```
`response` can be an `Exception` type and IMHO it is not ideal to compare an Exception to a bool

alizbay94 replied:
The `terminate_resource()` method has to return a string containing the error message if there is an error so that message can be included in the slack output, but I suppose it doesn't necessarily have to return a boolean `True` if the termination is successful. I we wish to keep the return types of the terminate method the same regardless of success or failure, we could return the string 'true' if the termination is successful rather than a Boolean, so that the return value is always a string. If success and failure both resulted in a string, the logic could be changed to `if response == "true"`; however, I prefer to leave it as is and return a boolean `True' upon success, and a string with the error message upon failure - I don't anticipate any issues that could arise from the comparison in this case, but if there could be negative implications from the comparison i'm overlooking, I can make the change.

CodyRay replied:
I agree that Exception and bool are not a great combo to return, i'd suggest using `None/String` or a list of errors. 

You can also effectively return two values in python

```
def terminate():
  return True, error_str

success, errors = terminate()
```

alizbay94 replied:
I like the idea of using a `None` / `String` combo as return values.

A list of errors would make sense for the AMI `terminate()` method, while a None / String combo would make more sense for the other resources since there would only be one error in those cases. 

In order to keep the return value of the `terminate()` method of each resource the same, i'll use the None / String return values for them all.

alizbay94 committed:
Remove return type from methods, and return exceptions as strings

CodyRay commented on app/ami.py:
```diff
@@ -134,8 +134,8 @@ def terminate_resource(self, dryrun: bool) -> bool:
-                    snapshots_deleted = False  # set to False and continue attempting to delete remaining Snapshots
-        return snapshots_deleted
+                    errors += f"{str(e)}\n"
+        return errors if errors.strip() else True
```
Instead of returning true or a string, you might consider returning a list of errors. If the list is empty then the caller knows that the call must have succeeded without any errors

CodyRay commented on app/nagbot.py:
```diff
@@ -116,11 +116,15 @@ def execute_internal(channel, dryrun):
+                        contact = sqslack.lookup_user_by_email(r.contact)
+                        message = message + r.make_resource_summary() + \
+                            f', "Terminate after"={r.terminate_after}, "Monthly Price"=' \
+                            f'{util.money_to_string(r.monthly_price)}, Contact={contact}\n'
```
You might consider keeping this in the case of failure, it could be handy to have the contact there to help poke 'em

sejal-raval reviewed (approved):
None/String combination for return values seems a better approach to me too. 
Overall code LGTM.

alizbay94 committed:
Use string / none combo as return values for terminating resources, and include typing Union to provide multiple return values as possibility for terminate methods

alizbay94 commented:
Updated the `terminate` methods to return `None` if there are no errors, and a `str` with errors if errors occur during termination.
Adjusted pytests for the new setup of the terminate methods.
Used `Union` from the `typing` module to include a typehint for the `terminate` methods indicating a possibility of returning a string or None.

alizbay94 committed:
Increment NagBot version
